### DISCUSSION TITLE
Lost notifications from BLE peripheral

### DISCUSSION THREAD

**Discussion by swiergot** [micropython/micropython#10050] [Using MicroPython]
I have a ESP32 board that is connecting to a LYWSD03MMC (with [this custom firmware](https://github.com/pvvx/ATC_MiThermometer/)) temperature & humidity sensor in order to get a history of readings. It's basically doing the same as [this webpage](https://pvvx.github.io/ATC_MiThermometer/TelinkMiFlasher.html) which uses the Web Bluetooth API.

Briefly speaking, I'm subscribing to a characteristic and then write to it a request for the last 50 readings. And in general it is working as I'm getting the readings via `IRQ_GATTC_NOTIFY`. The problem is that I should be getting 51 notifications (as the characteristic's value needs to change that many times in order to transfer all 50 readings plus 1 change to indicate it's done) but instead I'm getting only 15-20 notifications (different number each time). Each packet has a sequence number from 1 to 50 in the payload so I can see that the sensor gets the request right and sends all 50 readings, but simply some of the notifications are missing (there are gaps in the sequence numbers).

My question: is it at all possible that the notifications are lost within the ESP32? I think they are sent by the sensor one after another so they arrive fast. I'd expect that they queue up in some ESP32 buffers and the IRQ handler is called for every single one of them.

It works fine with the Web BT API webpage, all are received, so I assume the sensor works fine.

I should also mention that I tried to subscribe for indications but I got `IRQ_GATTC_NOTIFY`, like the sensor does not support indications and always sends notifications.

Any ideas?

Thanks!

**Comment by puppet13th** [comment]
have you tried using [aioble](https://github.com/micropython/micropython-lib/tree/master/micropython/bluetooth/aioble) module?

**Comment by swiergot** [comment]
Yes, it's the same. Also tried Arduino/ESP32 and it's working perfectly.

**Reply by jimmo** [reply]
@swiergot  Whcih version of MicroPython are you using? The way MicroPython does BLE on ESP32 was changed significantly (in a way that would improve exactly the sort of issue you're describing) _after_ v1.19.1 was released. So would be interested to see if you see the same problem on a nightly build.
